### ISSUE TITLE
TVH on Vero4k

### ISSUE THREAD

**Issue by BGI-YYZ** [edit4ever/script.module.zap2epg#42]
Hi, I'm having an issue with trying to get everything to work here.  I am no expert at this, so please excuse me in advance.  

I've tried setting up TVHeadend in various ways, but when I eventually try and run zap2epg, I get this error:  

`2021-07-30 14:38:34.022 spawn: Executing "/usr/bin/tv_grab_zap2epg"`
`2021-07-30 14:38:34.309 spawn: Traceback (most recent call last):`
`2021-07-30 14:38:34.309 spawn:   File "/home/osmc/.kodi/addons/script.module.zap2epg/zap2epg.py", line 17, in `
`2021-07-30 14:38:34.310 spawn:     import urllib.request, urllib.error, urllib.parse`
`2021-07-30 14:38:34.310 spawn: ImportError: No module named request`
`2021-07-30 14:38:34.361 xmltv: /usr/bin/tv_grab_zap2epg: htsmsg_xml_deserialize error Unknown syntatic element: `
`2021-07-30 14:38:34.361 xmltv: /usr/bin/tv_grab_zap2epg: grab returned no data`

This is using zap2epg v.2.0.3 on Kodi 19.  TVH is version 4.2.8

Everything seems to be working and I can watch TV, but the EPG screen is just blank.  

Thanks in advance.

**Comment by edit4ever** [comment]
Try uninstalling and reinstalling the zap2epg addon again.  It seems your system is not installing the additional python modules that are needed.

What OS system are you running (linux, LibreELEC, CoreELEC,etc)??

**Comment by BGI-YYZ** [comment]
I tried uninstalling and reinstalling, doesn't seem to have worked.

The device is a Vero 4k which is OSMC based on Linux.  If I type `uname -a` I get this:
`Linux Vero4k 4.9.113-44-osmc #1 SMP PREEMPT Fri Jun 18 22:55:31 UTC 2021 aarch64 GNU/Linux`

**Comment by edit4ever** [comment]
Strange - the issue is with OSMC installing the correct python version.

Just for the heck of it - try downloading the 1.31 version and see if it will install:
https://github.com/edit4ever/script.module.zap2epg/releases/tag/v1.3.1

Then if that installs - maybe you can install the current version.  I'm not sure why OSMC is not installing the python modules.
I may be able to look into that over the weekend.

**Comment by BGI-YYZ** [comment]
No dice on version 1.31, I get the following error:

`Failed to install Add-on from zip file`
`Dependency on xbmc.python version 2.7.13 could not be satisfied.`

I'll ask on the osmc forum and see if they have any ideas.  It had been working up to about a month ago.  I didn't do much with it, but noticed you had a version 2 of zap2epg and figured that would possibly fix it.  I was told I had to use Kodi 19, so I upgraded it from 18 and, well, here we are.

**Comment by BGI-YYZ** [comment]
As a follow up, someone on the OSMC forum suggested going through and removing all orphaned dependencies, which I did.

Going back to install zap2epg 2.0.3 gives me the error of
xbmcswift2, installation failed.  

The search continues

**Comment by edit4ever** [comment]
I probably should look into replacing/removing xbmcswift2 - but it should be compatible with Py3 and therefore Kodi 19.

It runs on LibreELEC 10 - so not sure what's missing in OSMC.

**Comment by edit4ever** [comment]
BTW - here is the addon for OSMC - so not sure why it didn't install:

https://mirrors.mit.edu/osmc/osmc/download/kodi/addons/matrix/script.module.xbmcswift2/

**Comment by edit4ever** [comment]
If you want - try this version - I changed the import xbmcswift2 line to lowercase plugin...maybe that's tripping up OSMC

https://1drv.ms/u/s!AhL4aRs7Xd05i4Q0ELWDL8beD2riiw?e=bZIzfn

**Comment by BGI-YYZ** [comment]
I manually installed xbmswift2 from the link you provided and it appears to have worked.  When I run zip2epg it says it successfully ran.  I went to my computer and used WinSCP to log into `/home/osmc/.kodi/userdata/addon_data/scriptmodule.zap2epg` and opened up xmltv.xml.  There's a nice list of everything that should be there.  So, thank you for that.

Unfortunately, I can't get TVHeadend to use this information in any way.

The search continues.

**Comment by edit4ever** [comment]
Now log into the tvheadend web interface.  There under the CONFIGURATION > CHANNEL/EPG > EPG GRABBER MODULE tab you should see the zap2epg grabber.  Make sure it is enabled.  If so, click on the re-run internal epg grabber button.

Then go to the CONFIGURATION > CHANNEL/EPG > CHANNELS tab and make sure that the EPG source is assigned for each channel.  If it didn't automatically find them - then you can manually assign.  Usually it will automatically find them after a restart.

Once those are mapped - you can rerun the internal grabber again to force a refresh and when it finished - you should see guide data in the tvheadend webUI.  It may take a little while to load into OSMC or you may need one more restart to get it to display.

Hope that helps!

**Comment by BGI-YYZ** [comment]
Something is seriously wrong, though I don't think it necessarily is with zap2epg.  I have managed to install zap2epg on my Vero4k and, after setting it up, I run it, it says that it has successfully downloaded a number of channels and stuff.  All seems good.

Going into TVH web interface is another story.  In Configuration -> Channel/EPG -> EPG Grabber Modules, there is no Zap2epg listed, even though I correctly put the TVH address and login info in the zap2epg configuration page.  Rebooted the Vero4k, stopped and started the TVH service, etc.  

As I've mentioned, I've been trying to sort this out in the [OSMC forum](https://discourse.osmc.tv/t/vero4k-kodi-19-and-add-ons/89915) as well.

**Comment by edit4ever** [comment]
That usually indicates the issue with permissions - OSMC is not the same setup as LibreELEC.  So what you need to do is find the tv_grab_zap2epg file - usually located in usr/bin.  You may also need to copy it there directly.  
In any case - once you find it - you need to make sure it has executable permissions (especially for the tvheadend hts user - but you can set for all).  That should then make it available in tvheadend.

**Comment by wbeard52** [comment]
I have a Vero4K and am running the stable 18.9 version.  One thing I had to do was to put the zap3epg.py file in the /usr/bin folder and change permissions so that everyone can execute it. chown 777 or something similar.

Only after I did that would TVH see Ap2epg and import the data.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by BGI-YYZ** [comment]
I was asked to upload the Kodi logs, which I did at [https://paste.osmc.tv/izagavavej](https://paste.osmc.tv/izagavavej) - there are a few mentions of zap2epg failing, but plenty of others that it does seem to work.  For instance
`2021-08-02 12:50:46.196 T:10241   ERROR <general>: GetDirectory - Error getting plugin://script.module.zap2epg/channels`
`2021-08-02 12:50:46.196 T:10242    INFO <general>: CPythonInvoker(35, /home/osmc/.kodi/addons/script.module.zap2epg/default.py): script successfully run`
`2021-08-02 12:50:46.208 T:9666    ERROR <general>: CGUIMediaWindow::GetDirectory(plugin://script.module.zap2epg/channels) failed`

**Comment by edit4ever** [comment]
Make sure you have also installed the tv_grab_zap2epg file in /usr/bin and have set its permissions to executable.  That is the grabber for tvheadend.  It is a file that automatically runs zap2epg.

**Comment by BGI-YYZ** [comment]
Sorry for the delay.

@edit4ever tv_grab_zap2epg is indeed in the /usr/bin folder  and permissions are set to octal 0755, so it is executable.  

@wbeard52 I have found 2 copies of zap2epg.py on my Vero.  They are located in 

`./proc/self/task/29324/cwd/script.module.zap2epg` which is dated Oct 31, 2020 and `./proc/self/task/29324/cwd/.kodi/addons/script.module.zap2epg` which is dated August 2, 2021.  

I copied the most recent one into /usr/bin and set it's permissions to octal 0777.  Rebooted the device and zap2epg still does not appear on the list of grabbers in TVH.  This is all I see.

https://i.imgur.com/D0sNnBQ.png

I thought maybe I could use the 4th module down, but there doesn't appear to be a way to change the name of the xml file it uses.

**Comment by edit4ever** [comment]
That's weird.  I'll go ahead and get an OSMC install running on a raspberry pi I have and see if I have the same issue.

**Comment by edit4ever** [comment]
OK - this is weird - but you see the one labeled WARNING: You are using SSH...
If you click on it and look in the right panel in the tvheadend webui - you'll see that that is the zap2epg grabber.  Something is getting messed up when copying it to usr/bin in tvheadend.

However, you should be able to enable it and see if you can run the internal grabber.

**Comment by edit4ever** [comment]
BTW - I installed on my RPi3 (11-20 version) and I had to use zap2epg release 1.31 for the install.  Looks like OSMC hasn't fully switched to Python 3 or a full Kodi 19 release - so the Python 2 stuff is confusing the addon.

**Comment by wbeard52** [comment]
OSMC is finalizing a major update to a new kernel 4.9 and Kodi 19.1 hopefully.  It's been awhile since they have updated but you can download a development branch by changes the sources.lst to the osmc-devel.  There are instructions on their forum.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by edit4ever** [comment]
Thanks - I'm good with LibreELEC.  Did enabling the strangely named grabber work for you?

**Comment by BGI-YYZ** [comment]
@edit4ever 
I see what you mean about the EPG Grabber Module that says "Warning..."  It does seem to be the zap2epg module.  So, I enabled it, disabled everything else and chose "re-run internal EPG Grabbers".  Unfortunately, it didn't work.  This is what the TVH log said about it:
`2021-08-05 12:41:23.401 xmltv: /usr/bin/tv_grab_zap2epg: grab /usr/bin/tv_grab_zap2epg`
`2021-08-05 12:41:23.403 spawn: Executing "/usr/bin/tv_grab_zap2epg"`
`2021-08-05 12:41:23.673 spawn: Traceback (most recent call last):`
`2021-08-05 12:41:23.673 spawn:   File "/home/osmc/.kodi/addons/script.module.zap2epg/zap2epg.py", line 17, in `
`2021-08-05 12:41:23.673 spawn:     import urllib.request, urllib.error, urllib.parse`
`2021-08-05 12:41:23.673 spawn: ImportError: No module named request`
`2021-08-05 12:41:23.705 xmltv: /usr/bin/tv_grab_zap2epg: htsmsg_xml_deserialize error Unknown syntatic element: `
`2021-08-05 12:41:23.705 xmltv: /usr/bin/tv_grab_zap2epg: grab returned no data`

**Comment by edit4ever** [comment]
The issue here is that the version of OSMC that you are running is not ready to be used completely.  It seems that they haven't fully finished the Python 3 updates.

Perhaps you can try to manually install the missing python modules?

Otherwise - if you go back to the 11.20 version of OSMC - you should be able to install the 1.3.1 version of zap2epg.

**Comment by BGI-YYZ** [comment]
I see.  I'm not going to go back to an old version of OSMC, but would be willing to try manually installing the missing python modules.  Errr, how exactly would I go about doing that?  Can you tell what modules are missing from my TVH error log?

**Comment by edit4ever** [comment]
Go ahead and try this update:
https://1drv.ms/u/s!AhL4aRs7Xd05i4Q6_ejMVa_lHbPS0Q?e=gsPsmc

Let me know how it goes!

**Comment by BGI-YYZ** [comment]
I'm sorry, can this file be made public?

**Comment by edit4ever** [comment]
? That is a public link to the file - are you getting an error when trying to download?

**Comment by BGI-YYZ** [comment]
I'm sorry, when trying to download it using my tablet, it insisted that I sign up for a Microsoft account.  Doing it from my computer worked fine.  I'll let you know how I get on.

**Comment by BGI-YYZ** [comment]
Okay, I uninstalled the previous zap2epg and installed the 2.04 and installed 2.04a.  It seemed to install okay, but when I went to run it, it gave an error.  From the log, it appears to be this:

`2021-08-06 14:51:35.406 T:16780    INFO <general>: initializing python engine.`
`2021-08-06 14:51:38.354 T:16780   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--`
                                                  
  - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'NameError'>
                                                   Error Contents: name 'Plugin' is not defined
                                                   Traceback (most recent call last):
                                                     File "/home/osmc/.kodi/addons/script.module.zap2epg/default.py", line 42, in <module>
                                                       plugin = Plugin()
                                                   NameError: name 'Plugin' is not defined

                                                   -->End of Python script error report<--
                                                   
`2021-08-06 14:51:38.930 T:16780    INFO <general>: Python interpreter stopped`
`2021-08-06 14:51:38.947 T:16778   ERROR <general>: GetDirectory - Error getting plugin://script.module.zap2epg/`
`2021-08-06 14:51:38.954 T:2592    ERROR <general>: CGUIMediaWindow::GetDirectory(plugin://script.module.zap2epg/) failed`

You can view the entire log at [https://paste.osmc.tv/fuqovabequ](https://paste.osmc.tv/fuqovabequ)

**Comment by BGI-YYZ** [comment]
It seems like zap2epg is sort of working.  I still get errors and the EPG is empty, but TVHeadend seems to be aware that it is there.  The TVH log says the following when I choose Configuration -> Channel/EPG -> EPG Grabber Modules -> Re-run Internal EPG Grabbers.

2021-08-19 11:29:05.399 xmltv: /usr/bin/tv_grab_zap2epg: grab /usr/bin/tv_grab_zap2epg
2021-08-19 11:29:05.402 spawn: Executing "/usr/bin/tv_grab_zap2epg"
2021-08-19 11:29:05.665 spawn: Traceback (most recent call last):
2021-08-19 11:29:05.665 spawn:   File "/home/osmc/.kodi/addons/script.module.zap2epg/zap2epg.py", line 17, in 
2021-08-19 11:29:05.665 spawn:     import urllib.request, urllib.error, urllib.parse
2021-08-19 11:29:05.665 spawn: ImportError: No module named request
2021-08-19 11:29:05.767 xmltv: /usr/bin/tv_grab_zap2epg: grab took 0 seconds
2021-08-19 11:29:05.797 xmltv: /usr/bin/tv_grab_zap2epg: parse took 0 seconds
2021-08-19 11:29:05.797 xmltv: /usr/bin/tv_grab_zap2epg:  channels   tot=   94 new=    0 mod=    0
2021-08-19 11:29:05.797 xmltv: /usr/bin/tv_grab_zap2epg:  brands     tot=    0 new=    0 mod=    0
2021-08-19 11:29:05.797 xmltv: /usr/bin/tv_grab_zap2epg:  seasons    tot=    0 new=    0 mod=    0
2021-08-19 11:29:05.797 xmltv: /usr/bin/tv_grab_zap2epg:  episodes   tot=    0 new=    0 mod=    0
2021-08-19 11:29:05.797 xmltv: /usr/bin/tv_grab_zap2epg:  broadcasts tot=    0 new=    0 mod=    0

I really don't want to, but am trying to figure out if I should be wiping out my device and starting again.

**Comment by edit4ever** [comment]
That may have been my mistake - I missed a change in the default.py file.

Try this version: https://1drv.ms/u/s!AhL4aRs7Xd05i4RS1GfWeHHPZZdybg?e=LOYm9E

Thanks!

**Comment by BGI-YYZ** [comment]
If I understand what you mean by the epg grabber channels set to zap2epg, if I go to Configuration -> Channel/EPG -> EPG Grabber Channels, all of the stations in the list, under the "path" column, point to usr/bin/tv_grab_zap2epg

Anyway, after updating, I get different errors.  Namely:

2021-08-19 12:04:00.000 xmltv: /usr/bin/tv_grab_zap2epg: grab /usr/bin/tv_grab_zap2epg
2021-08-19 12:04:00.002 spawn: Executing "/usr/bin/tv_grab_zap2epg"
2021-08-19 12:04:00.269 spawn: Traceback (most recent call last):
2021-08-19 12:04:00.269 spawn:   File "/home/osmc/.kodi/addons/script.module.zap2epg/zap2epg.py", line 17, in 
2021-08-19 12:04:00.269 spawn:     import urllib.request, urllib.error, urllib.parse
2021-08-19 12:04:00.270 spawn: ImportError: No module named request
2021-08-19 12:04:00.373 xmltv: /usr/bin/tv_grab_zap2epg: grab took 0 seconds
2021-08-19 12:04:00.403 xmltv: /usr/bin/tv_grab_zap2epg: parse took 0 seconds
2021-08-19 12:04:00.403 xmltv: /usr/bin/tv_grab_zap2epg:  channels   tot=   94 new=    0 mod=    0
2021-08-19 12:04:00.403 xmltv: /usr/bin/tv_grab_zap2epg:  brands     tot=    0 new=    0 mod=    0
2021-08-19 12:04:00.403 xmltv: /usr/bin/tv_grab_zap2epg:  seasons    tot=    0 new=    0 mod=    0
2021-08-19 12:04:00.403 xmltv: /usr/bin/tv_grab_zap2epg:  episodes   tot=    0 new=    0 mod=    0
2021-08-19 12:04:00.403 xmltv: /usr/bin/tv_grab_zap2epg:  broadcasts tot=    0 new=    0 mod=    0
2021-08-19 12:22:35.433 xmltv: /usr/bin/tv_grab_zap2epg: grab /usr/bin/tv_grab_zap2epg
2021-08-19 12:22:35.436 spawn: Executing "/usr/bin/tv_grab_zap2epg"
2021-08-19 12:22:35.766 spawn: Traceback (most recent call last):
2021-08-19 12:22:35.766 spawn:   File "/home/osmc/.kodi/addons/script.module.zap2epg/zap2epg.py", line 814, in 
2021-08-19 12:22:35.767 spawn:     mainRun(userdata)
2021-08-19 12:22:35.767 spawn:   File "/home/osmc/.kodi/addons/script.module.zap2epg/zap2epg.py", line 38, in mainRun
2021-08-19 12:22:35.767 spawn:     settings = ET.parse(settingsFile)
2021-08-19 12:22:35.767 spawn:   File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1182, in parse
2021-08-19 12:22:35.769 spawn:     tree.parse(source, parser)
2021-08-19 12:22:35.769 spawn:   File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 647, in parse
2021-08-19 12:22:35.769 spawn:     source = open(source, "rb")
2021-08-19 12:22:35.769 spawn: IOError: [Errno 2] No such file or directory: '/home/osmc/.kodi/userdata/addon_data/script.module.zap2epg/settings.xml'
2021-08-19 12:22:35.793 spawn: cat: /home/osmc/.kodi/userdata/addon_data/script.module.zap2epg/xmltv.xml: No such file or directory
2021-08-19 12:22:35.794 xmltv: /usr/bin/tv_grab_zap2epg: no output detected
2021-08-19 12:22:35.794 xmltv: /usr/bin/tv_grab_zap2epg: grab returned no data

**Comment by edit4ever** [comment]
OK - here's the other weird thing - you're still seeing python 2.7 instead of python 3.  You're system is getting confused on which python modules to run.

Does the Vero OSMC version allow you to uninstall python 2.7 and only have python 3 installed?

As you can see in the log - the the current failing issue is the missing ElementTree module.  I don't know why this would be missing as it is a standard python module that should be in the python 2.7 installation.  OSMC must be customizing their python installs with only the modules they feel are necessary.  --Perhaps someone in the OSMC space know how to add that module?

This missing module is why the episodes can't be parsed - which is causing the failure.

**Comment by BGI-YYZ** [comment]
Okay, that's good info - I'll see if I can get any help from the OSMC community and let you know.  

Thanks for all your help!

**Comment by BGI-YYZ** [comment]
How do I completely uninstall zap2epg?  I uninstalled it through the Kodi interface, and said yes to deleting all settings and rebooted the device. However, zap2epg is still listed as an available EPG Grabber Module in TVH. I would have expected it to disappear from the list.

**Comment by BGI-YYZ** [comment]
From the OSMC forum, it was suggested that I edit
`script.module.zap2epg/bin/tv_grab_zap2epg`
and change line 28 from
`python $ADDON_DIR/zap2epg.py`
to 
`python3 $ADDON_DIR/zap2epg.py`

All of a sudden, TVH's EPG is populating as it should.

Thank you for your help!

**Comment by edit4ever** [comment]
OK - that makes sense since they have both Python versions installed.  So you have to directly tell the system to use Python 3.

They should eventually remove Python 2 as it is at end of life.

But glad it's working for you!

Since that is working - you can remove the tv_grab_zap2epg file that you copied to usr/bin if you want as well.

**Comment by edit4ever** [comment]
It looks like your system (OSMC) may have copied that file to a few other directories.  Also possible that you copied one to the usr/bin during testing.  You can manually delete all of the ones other than the one in .kodi/addons/script.module.zap2epg/bin

Deleting zap2epg from within the Kodi interface (which should be the same for within the OSMC interface) will remove the files in the  .kodi/addons/script.module.zap2epg and .kodi/userdata/addon_data/script.module.zap2epg folders.

Anything else you find on the system - wasn't placed there by the zap2epg install - so feel free to delete.

Hope that helps.

**Comment by BGI-YYZ** [comment]
I'm so sorry to keep reopening this thread.  I went and deleted everything zap2epg related and installed the python3 version you made for me. 

> https://github.com/edit4ever/script.module.zap2epg/issues/42#issuecomment-906727610


It installs without issue, but when I go to run it, I get this:

2021-08-29 10:49:32.220 T:28746   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'NameError'>
                                                   Error Contents: name 'Plugin' is not defined
                                                   Traceback (most recent call last):
                                                     File "/home/osmc/.kodi/addons/script.module.zap2epg/default.py", line 42, in <module>
                                                       plugin = Plugin()
                                                   NameError: name 'Plugin' is not defined
                                                   -->End of Python script error report<--

**Comment by edit4ever** [comment]
So that issue is due to the OSMC team either not updating xbmcswift to a python 3 compatible version or not having it available in their python modules.

In all actuality - the xbmcswift use was an easy workaround when I first built the addon as I didn't have time to code in my own xbmc windows.  Since it always worked, I never went back to work around it.  Perhaps someone on the OSMC forum has a solution to get xbmcswift running?

Also - since this addon was designed to be used under LibreELEC - I follow(ed) their system guidelines of what is available for addons to use.  OSMC has their own reasons to only make certain modules available, so it leads to incompatibility.  The core zap2epg.py program will run as will the tv_grab_zap2epg grabber for tvheadend - the error your getting is the addon interface for zap2epg as it tries to load the addon main menu.
